File: frontend/src/lib/recruiter-talent-pool.ts
export const RECRUITER_TALENT_POOL_ROUTE = "/recruiter/talent-pool";
export const RECRUITER_TALENT_POOL_IMPORT_ROUTE = "/recruiter/talent-pool/import";

export const RECRUITER_TALENT_POOL_MARKERS = {
  page: "recruiter-talent-pool-page",
  summaryPanel: "recruiter-talent-pool-summary-panel",
  qualityPanel: "recruiter-talent-pool-quality-panel",
  sourceCoverage: "recruiter-talent-pool-source-coverage",
  recordsList: "recruiter-talent-pool-records-list",
  emptyState: "recruiter-talent-pool-empty",
  importLink: "recruiter-talent-pool-import-link",
  importPage: "recruiter-talent-pool-import-page",
  csvPaste: "recruiter-talent-pool-csv-paste",
  previewPanel: "recruiter-talent-pool-preview-panel",
  commitButton: "recruiter-talent-pool-commit-button",
  resultPanel: "recruiter-talent-pool-result-panel",
} as const;

export const RECRUITER_TALENT_POOL_FORBIDDEN_PII = [
  "email",
  "phone",
  "cv_text",
  "cv_raw",
  "phone_number",
  "@",
  "linkedin",
] as const;

export const TALENT_POOL_CSV_TEMPLATE = `display_name,job_title,skills,location,seniority,external_ats_id,candidate_id,pipeline_status
Alex Kowalski,Senior Backend Engineer,Python;FastAPI;PostgreSQL,Warsaw,senior,ATS-1001,CAND-42,review
`;

export type TalentPoolDataQuality = {
  score?: number;
  level?: "high" | "medium" | "low";
  warnings?: string[];
};

export type TalentPoolRecord = {
  id: number;
  display_name: string;
  job_title?: string | null;
  location?: string | null;
  seniority?: string | null;
  skills?: string[];
  data_quality?: TalentPoolDataQuality;
  external_ats_id?: string | null;
  candidate_id?: string | null;
  application_id?: number | null;
  pipeline_status?: string | null;
  source?: string;
  created_at?: string | null;
};

export type TalentPoolSummary = {
  total_records: number;
  shown: number;
  quality_high: number;
  quality_medium: number;
  quality_low: number;
  import_batches: number;
  last_import_at?: string | null;
  last_import_status?: string | null;
};

export type TalentPoolPayload = {
  company_slug: string;
  summary: TalentPoolSummary;
  data_quality: {
    levels: Record<string, number>;
    top_warnings: { code: string; count: number }[];
  };
  source_coverage: {
    imported_internal_pool: number;
    import_sources: Record<string, number>;
    external_sourcing: boolean;
    live_ats_sync: boolean;
  };
  items: TalentPoolRecord[];
  scope_note: string;
};

export type TalentPoolPreviewRow = {
  row_index: number;
  display_name: string;
  job_title?: string | null;
  location?: string | null;
  skills?: string[];
  status: "ready" | "duplicate" | "error";
  data_quality?: TalentPoolDataQuality;
  duplicate_key?: string;
};

export type TalentPoolPreviewPayload = {
  import_id: number;
  company_slug: string;
  status: string;
  summary: { total: number; ready: number; duplicates: number; errors: number };
  warnings: string[];
  rows: TalentPoolPreviewRow[];
};

export type TalentPoolCommitPayload = {
  import_id: number;
  company_slug: string;
  status: string;
  summary: { accepted: number; duplicates_skipped: number; total_previewed: number };
};

export function talentPoolRowHasForbiddenPii(row: Record<string, unknown>): boolean {
  const blob = JSON.stringify(row).toLowerCase();
  return RECRUITER_TALENT_POOL_FORBIDDEN_PII.some((marker) => blob.includes(marker));
}

export function talentPoolImportQuery(token: string, companySlug: string): string {
  return new URLSearchParams({ token, company_slug: companySlug }).toString();
}

File: backend/app/services/recruiter_talent_pool.py
"""Recruiter talent pool — list, summarize quality, source coverage."""

from __future__ import annotations

import json
from collections import Counter
from typing import Any

from sqlalchemy.orm import Session

from app.database.models import RecruiterTalentPoolImport, RecruiterTalentPoolRecord
from app.services.recruiter_inbox import _require_company_slug


def _parse_json(raw: str | None, fallback: Any) -> Any:
    if not raw:
        return fallback
    try:
        return json.loads(raw)
    except json.JSONDecodeError:
        return fallback


def build_recruiter_talent_pool(
    db: Session,
    *,
    company_slug: str,
    locale: str = "en",
    limit: int = 50,
) -> dict[str, Any]:
    slug = _require_company_slug(company_slug)
    pl = locale.lower().startswith("pl")
    cap = max(1, min(limit, 100))

    records = (
        db.query(RecruiterTalentPoolRecord)
        .filter(
            RecruiterTalentPoolRecord.company_slug == slug,
            RecruiterTalentPoolRecord.archived_at.is_(None),
        )
        .order_by(RecruiterTalentPoolRecord.created_at.desc())
        .limit(cap)
        .all()
    )
    total = (
        db.query(RecruiterTalentPoolRecord)
        .filter(
            RecruiterTalentPoolRecord.company_slug == slug,
            RecruiterTalentPoolRecord.archived_at.is_(None),
        )
        .count()
    )
    imports = (
        db.query(RecruiterTalentPoolImport)
        .filter(RecruiterTalentPoolImport.company_slug == slug)
        .order_by(RecruiterTalentPoolImport.created_at.desc())
        .limit(10)
        .all()
    )

    quality_levels: Counter[str] = Counter()
    source_counts: Counter[str] = Counter()
    warning_counts: Counter[str] = Counter()
    items: list[dict[str, Any]] = []

    for rec in records:
        quality = _parse_json(rec.data_quality_json, {})
        level = str(quality.get("level") or "unknown")
        quality_levels[level] += 1
        for w in quality.get("warnings") or []:
            warning_counts[str(w)] += 1
        skills = _parse_json(rec.skills_json, [])
        source_counts["imported_internal_pool"] += 1
        items.append(
            {
                "id": rec.id,
                "display_name": rec.display_name,
                "job_title": rec.job_title,
                "location": rec.location,
                "seniority": rec.seniority,
                "skills": skills if isinstance(skills, list) else [],
                "data_quality": quality,
                "external_ats_id": rec.external_ats_id,
                "candidate_id": rec.candidate_id,
                "application_id": rec.application_id,
                "pipeline_status": rec.pipeline_status,
                "source": rec.source_type or "imported_internal_pool",
                "source_type": rec.source_type or "csv_import",
                "consent_visibility": rec.consent_visibility or "unknown",
                "archived": rec.archived_at is not None,
                "created_at": rec.created_at.isoformat() if rec.created_at else None,
            }
        )

    import_sources = Counter(i.import_source for i in imports)
    last_import = imports[0] if imports else None

    return {
        "company_slug": slug,
        "summary": {
            "total_records": total,
            "shown": len(items),
            "quality_high": quality_levels.get("high", 0),
            "quality_medium": quality_levels.get("medium", 0),
            "quality_low": quality_levels.get("low", 0),
            "import_batches": len(imports),
            "last_import_at": last_import.created_at.isoformat() if last_import and last_import.created_at else None,
            "last_import_status": last_import.status if last_import else None,
        },
        "data_quality": {
            "levels": dict(quality_levels),
            "top_warnings": [{"code": k, "count": v} for k, v in warning_counts.most_common(5)],
        },
        "source_coverage": {
            "imported_internal_pool": total,
            "import_sources": dict(import_sources),
            "external_sourcing": False,
            "live_ats_sync": False,
        },
        "items": items,
        "scope_note": (
            "Wewnętrzny pool strukturalny — import CSV, bez live sync ATS."
            if pl
            else "Structured internal pool — CSV import, no live ATS sync."
        ),
    }


def list_talent_pool_records_for_radar(
    db: Session,
    *,
    company_slug: str,
    limit: int = 50,
) -> list[RecruiterTalentPoolRecord]:
    slug = _require_company_slug(company_slug)
    return (
        db.query(RecruiterTalentPoolRecord)
        .filter(RecruiterTalentPoolRecord.company_slug == slug)
        .order_by(RecruiterTalentPoolRecord.created_at.desc())
        .limit(limit)
        .all()
    )
